### PR TITLE
address some metadata hashtree issues

### LINKED ISSUE #389: Metadata Hashtree should *not* trap exits
instead it should spawn and monitor. see https://github.com/basho/riak_core/pull/385#discussion_r6504153

### PR DESCRIPTION
**Pull request by jrwest**
fixes #389 #390 #391. see issues+commits for more details

### CODE REVIEW

File: src/riak_core_broadcast.erl [hunk 1]
```diff
@@ -222,7 +246,12 @@ handle_call({get_peers, Root}, _From, State) ->
     LazyPeers = all_peers(Root, State#state.lazy_sets, State#state.common_lazys),
     {reply, {EagerPeers, LazyPeers}, State};
 handle_call(broadcast_members, _From, State=#state{all_members=AllMembers}) ->
-    {reply, AllMembers, State}.
+    {reply, AllMembers, State};
+handle_call(exchanges, _From, State=#state{exchanges=Exchanges}) ->
+    {reply, Exchanges, State};
+handle_call({cancel_exchanges, WhichExchanges}, _From, State) ->
+    Cancelled = cancel_exchanges(WhichExchanges, State#state.exchanges),
+    {reply, Cancelled, State}.
```
Comment (lenary) [inline src/riak_core_broadcast.erl:254]: Do you not want to update the `State` here to remove the exchanges that you've just cancelled?

Comment (jrwest) [inline src/riak_core_broadcast.erl:254]: They will be cleaned up when their monitor fires https://github.com/basho/riak_core/blob/jrw-metadata-hashtree-updates/src/riak_core_broadcast.erl#L307-L309

**Comment by lenary** [comment]
Looks good to me :+1:

File: src/hashtree_tree.erl [hunk 2]
```diff
@@ -237,9 +237,9 @@ update_snapshot(Tree=#hashtree_tree{dirty=Dirty,nodes=Nodes,snapshot=Snapshot0})
     catch ets:delete(Snapshot0),
     FoldRes = gb_sets:fold(fun(DirtyName, Acc) ->
                                    DirtyKey = node_key(DirtyName, Tree),
-                                   DirtyNode = lookup_node(DirtyName, Tree),
-                                   {SnapNode, NewNode} = hashtree:update_snapshot(DirtyNode),
-                                   [{{DirtyKey, SnapNode}, {DirtyKey, NewNode}} | Acc]
+                                   Node = lookup_node(DirtyName, Tree),
+                                   {DirtyNode, NewNode} = hashtree:update_snapshot(Node),
+                                   [{{DirtyKey, DirtyNode}, {DirtyKey, NewNode}} | Acc]
```
Comment (engelsanchez) [inline src/hashtree_tree.erl:242]: I'm getting a feeling you like it dirty...

Comment (jrwest) [inline src/hashtree_tree.erl:242]: hahaha. it happened to be the smaller change to make

File: src/riak_core_broadcast.erl [hunk 3]
```diff
@@ -400,6 +429,43 @@ exchange(Peer, State=#state{mods=[Mod | Mods],exchanges=Exchanges}) ->
              end,
     State1#state{mods=Mods ++ [Mod]}.
 
+cancel_exchanges(all, Exchanges) ->
+    kill_exchanges(Exchanges);
+cancel_exchanges(WhichProc, Exchanges) when is_reference(WhichProc) orelse
+                                            is_pid(WhichProc) ->
+    KeyPos = case is_reference(WhichProc) of
+              true -> 3;
```
Comment (engelsanchez) [inline src/riak_core_broadcast.erl:437]: BTW, this is the kind of thing that makes you go "Hmmm, this should be a record".

**Comment by engelsanchez** [comment]
I have reviewed the code and exercised it manually by hacking a long exchange and issuing exchange listing and cancellation commands, and by redbugging the riak_core_metadata_hashtree and watching it handle monitor events.

I have a note about considering using records for exchanges. That would be a nice cleanup that would remove some magic constants at least.  There is the caveat that tests for this code are scheduled to be written later, so I'm trusting @jrwest to give these modules good coverage when the time comes :).

:+1: :dancer: :boat:

**Pull request merged**